subprojects/core/src/main/java/org/gradle/caching/internal/tasks/TarTaskOutputPacker.java

```diff
@@ -51,23 +50,33 @@
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
+import java.nio.file.Files;
 import java.util.Collection;
 import java.util.Map;
 import java.util.SortedSet;
-import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
+import static org.gradle.caching.internal.tasks.TaskOutputPackerUtils.ensureDirectoryForProperty;
+import static org.gradle.caching.internal.tasks.TaskOutputPackerUtils.makeDirectory;
+
 /**
- * Packages task output to a POSIX TAR file. Because Ant's TAR implementation
- * supports only 1 second precision for file modification times, we encode the
- * fractional nanoseconds into the group ID of the file.
+ * Packages task output to a POSIX TAR file.
  */
+@SuppressWarnings("Since15")
 public class TarTaskOutputPacker implements TaskOutputPacker {
     private static final String METADATA_PATH = "METADATA";
     private static final Pattern PROPERTY_PATH = Pattern.compile("(missing-)?property-([^/]+)(?:/(.*))?");
-    private static final long NANOS_PER_SECOND = TimeUnit.SECONDS.toNanos(1);
+    @SuppressWarnings("OctalInteger")
+    private static final int FILE_PERMISSION_MASK = 0777;
+    private static final int BUFFER_SIZE = 64 * 1024;
+    private static final ThreadLocal<byte[]> COPY_BUFFERS = new ThreadLocal<byte[]>() {
```

Comment:
What's the scope of this class? If it is long lived, then this is a memory leak.